### PR TITLE
Add support for calculating losses on noise and dereverberated signals

### PR DESCRIPTION
**Pull request by Emrys365**
This PR restores the support for calculating losses on noise and dereverberated signals.
The function has bee removed since the commit https://github.com/espnet/espnet/commit/2caec34f979b54410c9c9785812ecdf3d09ed6b8 due to the major refactorization of [espnet2/enh/espnet_model.py](https://github.com/espnet/espnet/blob/master/espnet2/enh/espnet_model.py).
This function is now re-implemented by adding two new attributes to the criterion module: `is_noise_loss` and `is_dereverb_loss`. They can also be easily configured in the yaml file.

Also, most speech enhancement/separation models provided in [espnet2/enh/separator/](https://github.com/espnet/espnet/blob/master/espnet2/enh/separator) have been modified to support predicting the noise signal in addition to the speech signals, i.e., the attribute `predict_noise` is added as an indicator.

### CODE REVIEW

**Comment by Emrys365** [comment]
@LiChenda Could you take a look at the changes?

**Comment by LiChenda** [comment]
> @LiChenda Could you take a look at the changes?

Sure!

File: espnet2/enh/loss/criterions/time_domain.py [hunk 1]
```diff
@@ -20,10 +20,37 @@ def name(self) -> str:
...
+        is_dereverb_loss=False,
+    ):
         super().__init__()
-        self._name = name
+        # only used during validation
         self._only_for_test = only_for_test
+        # only used to calculate the noise-related loss
+        self._is_noise_loss = is_noise_loss
+        # only used to calculate the dereverberation-related loss
+        self._is_dereverb_loss = is_dereverb_loss
```
Comment (LiChenda) [inline espnet2/enh/loss/criterions/time_domain.py:44]: I saw you added `only_for_test`, `is_noise_loss ` and `is_dereverb_loss ` for both time-domain loss and frequency-domain loss.  Adding these attributes to `AbsEnhLoss(torch.nn.Module, ABC)`  might be a more brief way. The current approach is good. It's just a suggestion.

File: espnet2/enh/separator/dc_crn_separator.py [hunk 2]
```diff
@@ -84,11 +86,13 @@ def __init__(
         super().__init__()
 
         self._num_spk = num_spk
+        self.predict_noise = predict_noise
         self.mode = mode
         if mode not in ("mapping", "masking"):
             raise ValueError("mode=%s is not supported" % mode)
         self.ref_channel = ref_channel
 
+        num_outputs = self.num_spk + 1 if self.predict_noise else self.num_spk
```
Comment (LiChenda) [inline espnet2/enh/separator/dc_crn_separator.py:95]: Given that all separators has the new attributes `num_outputs ` and `predict_noise `, maybe we should consider adding it to the super class `AbsSeparator`.

**Review by LiChenda — APPROVED** [review]
LGTM! Great Job! I just post some small suggestions!

File: espnet2/enh/separator/dc_crn_separator.py (hunk 2 above)
Comment (Emrys365) [reply espnet2/enh/separator/dc_crn_separator.py:95]: Not all models can easily support `predict_noise`, such as DPCL and DANet. So I prefer to add support for this function only when it needs minor modifications. Of course, we would recommend newly implemented models to try to support this function.

File: espnet2/enh/loss/criterions/time_domain.py (hunk 1 above)
Comment (Emrys365) [reply espnet2/enh/loss/criterions/time_domain.py:44]: My initial thought is that some losses may not be designed for noise signals, like the DPCL loss. So I didn't add these properties to `AbsEnhLoss` directly.
If you think it is better to inherit from `AbsEnhLoss`, I would do that.

**Comment by sw005320** [comment]
Thanks a lot!
